File: Web/Lab1/server/src/main/java/nothing/here/ResponseFactory.java
package nothing.here;

import java.nio.charset.StandardCharsets;

public class ResponseFactory {
    public static String createOkResponseRaw(String content, String content_type) {

        return """
            Content-Type: %s
            Content-Length: %d
            
            %s
            """.formatted(content_type, content.getBytes(StandardCharsets.UTF_8).length, content);
    }

    public static String createOkResponseText(String content) {
        return createOkResponseRaw(content, "text/plain");
    }

    public static String createOkResponseJson(String content) {
        return createOkResponseRaw(content, "application/json");
    }



    public static String createBadResponseErrorRaw(String content, String content_type) {
        content = content.replace("\"", "\\\"");
        content = "{ \"error\": \"" + content + "\" }";

        return """
            Status: 400 Bad Request
            Content-Type: %s
            Content-Length: %d
            
            %s
            """.formatted(content_type, content.getBytes(StandardCharsets.UTF_8).length, content);
    }

    public static String createBadResponseErrorText(String content) {
        return createBadResponseErrorRaw(content, "text/plain");
    }

    public static String createBadResponseErrorJson(String content) {
        return createBadResponseErrorRaw(content, "application/json");
    }


    public static String createInternalErrorResponseRaw(String content, String content_type) {
        content = content.replace("\"", "\\\"");
        content = "{ \"error\": \"" + content + "\" }";

        return """
            Status: 500 Internal Server Error
            Content-Type: %s
            Content-Length: %d
            
            %s
            """.formatted(content_type, content.getBytes(StandardCharsets.UTF_8).length, content);
    }

    public static String createInternalErrorResponseText(String content) {
        return createInternalErrorResponseRaw(content, "text/plain");
    }

    public static String createInternalErrorResponseJson(String content) {
        return createInternalErrorResponseRaw(content, "application/json");
    }
}

File: Web/Lab1-rusted/server/src/response_factory.rs
use multiline_str::multiline_str;

pub struct ResponseFactory;

impl ResponseFactory {
    pub fn ok_raw(content: &str, content_type: &str) -> String {
        format!(
            multiline_str!(
                "Content-Type: {}"; 
                "Content-Length: {}"; 
                ""; 
                "{}"
            ),
            content_type,
            content.len(),
            content
        )
    }

    #[allow(unused)]
    pub fn ok_text(text: &str) -> String {
        Self::ok_raw(text, "text/plain")
    }
    
    pub fn ok_json(text: &str) -> String {
        Self::ok_raw(text, "application/json")
    }
    
    pub fn bad_response_raw(content: &str, content_type: &str) -> String {
        format!(
            multiline_str!(
                "Status: 400 Bad Request";
                "Content-Type: {}"; 
                "Content-Length: {}"; 
                ""; 
                "{}"
            ),
            content_type,
            content.len(),
            content
        )
    }

    pub fn bad_response_text(text: &str) -> String {
        Self::bad_response_raw(text, "text/plain")
    }
    
    pub fn internal_error_raw(content: &str, content_type: &str) -> String {
        format!(
            multiline_str!(
                "Status: 500 Internal Server Error";
                "Content-Type: {}"; 
                "Content-Length: {}"; 
                ""; 
                "{}"
            ),
            content_type,
            content.len(),
            content
        )
    }

    pub fn internal_error_text(text: &str) -> String {
        Self::internal_error_raw(text, "text/plain")
    }
    
}
